### ISSUE TITLE
Thread deadlock when attempting to transmit data from Listener callback

### ISSUE THREAD

**Issue by pbleyer** [eclipse-cyclonedds/cyclonedds-python#7]
I am getting freezes to what it appears to be thread deadlocks when calling writers inside listener callbacks. In particular, after reading messages on a Listener `on_data_available` function I try to send reply messages by calling a writer that writes messages to the same topic. The calls to `writer.write` are able to send messages to the topic but afterward don't return.

Tested in Windows 10 x64 and Linux Ubuntu 20 with similar behavior. In Windows, the call stack for the offending thread looks like this:

```
0x0000000000000000
ntdll.dll!ZwWaitForAlertByThreadId+0x14
ntdll.dll!RtlSleepConditionVariableSRW+0x131
KERNELBASE.dll!SleepConditionVariableSRW+0x29
ddsc.dll!ddsrt_cond_wait+0x13
ddsc.dll!dds_reader_data_available_cb+0x5f
ddsc.dll!dds_wait_for_acks+0x40a
ddsc.dll!deadline_unregister_instance_real+0x194
ddsc.dll!dds_writecdr+0x30f
ddsc.dll!dds_write_flush+0x28a
ddsc.dll!dds_write+0x4a
ddspy.cp39-win_amd64.pyd!PyInit_ddspy+0x1f30
python39.dll!PyDict_MergeFromSeq2+0xbdd
python39.dll!PyEval_EvalFrameDefault+0x5382
python39.dll!PyDict_GetItem+0xbef
python39.dll!PyEval_EvalFrameDefault+0x3009
python39.dll!PyLong_AsDouble+0x723
python39.dll!PyEval_EvalFrameDefault+0x1df2
python39.dll!PyDict_GetItem+0xbef
python39.dll!PyEval_EvalFrameDefault+0x3009
python39.dll!PyEval_EvalFrameDefault+0x1d14
python39.dll!PyLong_AsDouble+0x723
python39.dll!PyEval_EvalFrameDefault+0x1df2
python39.dll!PyDict_GetItem+0xbef
python39.dll!PyFunction_Vectorcall+0x257
python39.dll!PyVectorcall_Call+0x5c
python39.dll!PyObject_Call+0x4f
_ctypes.pyd!DllGetClassObject+0x40e
_ctypes.pyd!DllGetClassObject+0x7f5
libffi-7.dll!ffi_call_go+0x28c
libffi-7.dll!ffi_prep_go_closure+0x19e
ddsc.dll!dds_reader_data_available_cb+0x160
ddsc.dll!dds_wait_for_acks+0x40a
ddsc.dll!deadline_unregister_instance_real+0x422
ddsc.dll!guid_prefix_zero+0x6606
ddsc.dll!guid_prefix_zero+0xb204
ddsc.dll!guid_prefix_zero+0x861a
ddsc.dll!guid_prefix_zero+0xc0f9
ddsc.dll!guid_prefix_zero+0x69dd
ddsc.dll!guid_prefix_zero+0xdb4c
ddsc.dll!create_thread_with_properties+0xdc
ddsc.dll!ddsrt_thread_setname+0x1dc
KERNEL32.DLL!BaseThreadInitThunk+0x14
ntdll.dll!RtlUserThreadStart+0x21
```

It zombies-out in `ddsrt_thread_setname`.

Is this the expected behavior? Are calls to writers disallowed in reader callbacks?

I am attaching a test that exhibits the issue. The `service.py` module provides a service function that waits for queries and sends back replies. The `echotest.py` module sends queries to the service and expects back "echo" replies to those messages. The service will try to send back the reply messages when processing the `on_data_available` callback and will consequently freeze.

[ddsrace.zip](https://github.com/eclipse-cyclonedds/cyclonedds-python/files/6544904/ddsrace.zip)

**Comment by thijsmie** [comment]
Hi @pbleyer,

After some review and a quick chat with @eboasson I think the issue here is that you generate a `on_data_available` callback within another `on_data_available` callback. Since listeners are called synchronously with a write on local delivery and protected by some mutexes this will deadlock. There is also the Python Global Interpreter lock in the mix which makes this a bit more complicated, but since it should be released and aquired when crossing the ctypes boundaries I think that is okay.

I see a few ways to go about doing this in a different way:
 * Switch to using a QueryTopic and a ReplyTopic. This avoids the same listener firing again in line with the first listener. It also might be more efficient in the long run, your Services that are not interested in the Replies just the Queries don't have to subscribe to them.
 * Avoid listener all together and use an async setup. You can use `async for message in reader.take_aiter()`.
 * Not yet supported, so more as a note for myself: Content filtered readers could also work in here but I don't know how they interact with readers.

In short: It is not disallowed to call writers in reader callbacks if those don't trigger another callback on the same entity.

**Comment by pbleyer** [comment]
Hi @thijsmie. Many thanks for your quick reply.

Makes sense. Further looking into the implementation I can see the mutex issues and that the callback will deadlock if reentrant.

I tried processing the messages in a separate thread but it looks like the locks or the GIL act before my code is executed. I also tried separate topics and was able to deadlock too, but I think that I had a bug in my previous implementation. Will check that again.

I was able to aiofy the processing using the aiter methods and that works without issues, at least for the service module.

It would be nice to have more filtering options. In my implementation my modules need to be promiscuous for monitoring. However I was reading the API to see if there is any way to filter-out the publishers or subscribers (e.g. by their id) but only found read and query conditions.

Many thanks again and kudos for the excellent work with the library.

**Comment by pbleyer** [comment]
Ok, for reference this is the implementation using coroutines.

The echo sender application is able to send and receives messages at a high rate and most of the CPU usage of the process is used by Python itself. However now I am seeing that if I use a small deadline time like 1ms, the service gets a dangling thread stuck at high CPU usage that remains even when I stop the echo app. 

Since I am stressing the communication, missing the deadlines is expected. But when I stop and restart the echo app several times while the service is running, it appears to get stuck in the dangling thread and becomes unresponsive.

In Windows, the stack of the dangling thread looks like this:
After starting the application, no messages yet:
```
ntdll.dll!ZwWaitForAlertByThreadId+0x14
ntdll.dll!RtlSleepConditionVariableSRW+0x131
KERNELBASE.dll!SleepConditionVariableSRW+0x29
ddsc.dll!ddsrt_cond_waitfor+0xb0
ddsc.dll!resched_xevent_if_earlier+0x7f4
ddsc.dll!create_thread_with_properties+0xdc
ddsc.dll!ddsrt_thread_setname+0x1dc
KERNEL32.DLL!BaseThreadInitThunk+0x14
ntdll.dll!RtlUserThreadStart+0x21
```
After sending high rate messages and stopping/restarting the echo app several times:
```
ntdll.dll!RtlReleaseSRWLockExclusive+0xa
ddsc.dll!dds_rhc_default_deadline_missed_cb+0xbb
ddsc.dll!deadline_unregister_instance_real+0x39
ddsc.dll!delete_xevent_callback+0xd4
ddsc.dll!resched_xevent_if_earlier+0x6ff
ddsc.dll!create_thread_with_properties+0xdc
ddsc.dll!ddsrt_thread_setname+0x1dc
KERNEL32.DLL!BaseThreadInitThunk+0x14
ntdll.dll!RtlUserThreadStart+0x21
```

I was expecting the service to recover gracefully from missed deadlines. Or is there something necessary to do or acknowledge when deadlines are missed and dds_rhc_default_deadline_missed_cb is called?

Looking at the source code of the libraries (https://github.com/eclipse-cyclonedds/cyclonedds/blob/46eda2d7786b55ca9704470b58c15cff3b93621f/src/core/ddsc/src/dds_rhc_default.c#L532) I don't think this has to do with Python but the CycloneDDS library, though. 

[ddsrace-async.zip](https://github.com/eclipse-cyclonedds/cyclonedds-python/files/6548905/ddsrace-async.zip)

**Comment by eboasson** [comment]
Thanks @pbleyer for not only reporting an interesting issue but also providing a useful reproducer. I haven't seen a deadlock or anything (my platform is different, so it may be that I got lucky), but I do see interesting things.

> I was expecting the service to recover gracefully from missed deadlines. Or is there something necessary to do or acknowledge when deadlines are missed and dds_rhc_default_deadline_missed_cb is called?

No, there is no need for any of that. The event should simply fire but have no other consequences. What you are seeing here is that the rate at which the event fires increases over time. Why is that?

Firstly, we have taken:

> For example, the value of the StatusChangedFlag associated with the REQUESTED_DEADLINE_MISSED status will become TRUE each time new deadline occurs (which increases the total_count field within RequestedDeadlineMissedStatus). The value changes to FALSE when the application accesses the status via the corresponding get_requested_deadline_missed_status method on the proper Entity.

from the DDS spec as meaning "the event is raised" and that therefore the intent is that the listener will be invoked periodically if no updates are published at all (I have this idea that it says so somewhere explicitly, but I can't find it right now, feel free to correct me!). Without that, the `total_count` field could be computed on requesting the status and there'd be no need for a missed deadline to induce any load if no-one actually looks at it.

Secondly, there is the assumption that a deadline miss will be a very rare event. I think this is in general a reasonable assumption: it is a notification of a failure to meet (hard) timing requirements, so presumably your system is either seriously broken by the time that happens, or it wasn't really a requirement to begin with. In the latter case: why ask for a notification?

The counter-argument to the combination of the above is of course that the deadline misses may be caused by a (CPU) overload condition and that inducing additional load in an overload situation is a blatantly stupid thing to do. The periodic setting of the status and invocation of the listener can probably be reworked and it is definitely worth considering that.

As it is, at the moment, the more deadline misses occur, the more often things get checked and the higher the load.

It turns out that the number of deadline misses grows because each write creates a new unique instance (= unique key value) and so no updates are ever published for any of these instances and all of them keep missing deadlines. So with a 1ms deadline, after the first write you have 1 deadline miss at 1kHz, then at 2 * 1kHz, then 3, &c. That blows up rather quickly.

The type you use is:
```python
@cdr
class Message:
        "Message object"
        import pycdr.types as ct
        # Message id (UUID) to identify requests and responses
        uid: ct.array[ct.uint8, 16]
...
```
which I would expect to either:
* result in a topic in which each field is a key. With the GUIDs and time stamps in there, you'd expect a unique instance for each write and the issues are: (1) using a deadline for non-periodic data, and (2) having a *huge* space leak because you are not "unregistering" the instance on the writer side;
* result in a "keyless" topic, with only a single instance. I believe this is the intent (right, @thijsmie?) as it fits with the program logic as well as with how IDL is generally used by the DDS spec.

I've checked, and the readers/writers created do indicate that their type is "keyless" because the last two (hex)digits of the writer's GUID are 03 (from running with `CYCLONEDDS_URI='<Tr><C>trace,rhc</>'`):
```
1622270636.277987 [0]    1777248: new_writer(guid abb41001:b6183716:578ef15a:103, (default).Message/comms::Message)
1622270636.277990 [0]    1777248: ref_participant(abb41001:b6183716:578ef15a:1c1 @ 0x101a043c0 <- abb41001:b6183716:578ef15a:103 @ 0x10330ab34) user 2 builtin 8
1622270636.277998 [0]    1777248: WRITER abb41001:b6183716:578ef15a:103 QOS={user_data=0<>,topic_name="Message",type_name="comms::Message",topic_data=0<>,group_data=0<>,durability=2,durability_service=0:0:1:-1:-1:-1,deadline=1000000,latency_budget=0,liveliness=0:9223372036854775807,reliability=0:1000000000,lifespan=9223372036854775807,destination_order=0,history=0:10,resource_limits=-1:-1:-1,ownership=0,ownership_strength=0,presentation=0:0:0,partition={},transport_priority=0,adlink_entity_factory=1,adlink_writer_data_lifecycle=1}
```
However:
```
1622270675.941832 [0]    1777248: write_sample abb41001:b6183716:578ef15a:103 #4: ST0 Message/comms::Message:
1622270675.941899 [0]    1777248:  => EVERYONE
1622270675.941915 [0]    1777248: rhc_store 95bdfbe864574cb4,95ea5c25111baa2 si 0 has_data 1: new instance
1622270675.941933 [0]    1777248: update_conditions_locked(0x103307fd0 0x101c05510) - inst 4 nonempty 1 disp 0 nowr 0 new 1 samples 1 read 0
...
1622270676.473863 [0]    1777248: write_sample abb41001:b6183716:578ef15a:103 #5: ST0 Message/comms::Message:
1622270676.473966 [0]    1777248:  => EVERYONE
1622270676.473980 [0]    1777248: rhc_store a1fa7fdd1ae929c0,95ea5c25111baa2 si 0 has_data 1: new instance
1622270676.473993 [0]    1777248: update_conditions_locked(0x103307fd0 0x1035044c0) - inst 5 nonempty 1 disp 0 nowr 0 new 1 samples 1 read 0
```
each sample results in a new instance (the first number following `rhc_store` is a unique id), the reader history cache says there are 4, then 5 instances when it processes the various attached conditions.

So clearly there's a discrepancy. If I change the type definition to
```python
@cdr(keylist=[])
class Message:
        "Message object"
        import pycdr.types as ct
        # Message id (UUID) to identify requests and responses
        uid: ct.array[ct.uint8, 16]
...
```

the problem goes away.

It looks to me that the underlying issue is that the python binding currently treats a plain `@cdr` as a keyless topic in one place and as a keyed topic in another place. @thijsmie, feel free to shoot down my theory!

**Comment by thijsmie** [comment]
Thank you for the in depth analysis @eboasson! Indeed you theory seems to have been correct. I have been interpreting 'keyless' wrong and that was reflected in the API. I have issued a PR !something that addresses this.

I changed the semantics as follow:
`@cdr` stands for "Keyed topic where all fields are key"
`@cdr(keyless=True)` stands for "Keyless topic, no fields are key"
`@cdr(keylist=[])` stands for "Keyed topic, no fields are key" which I don't know what the usecase would be.
